python_modules/dagster/dagster/_core/definitions/result.py

```diff
@@ -1,30 +1,35 @@
 from collections.abc import Mapping, Sequence
-from typing import NamedTuple, Optional
+from datetime import datetime  # noqa
+from os import PathLike  # noqa
+from typing import Generic, Optional, TypeVar
+
+from dagster_shared.check.decorator import checked
 
 import dagster._check as check
 from dagster._annotations import PublicAttr
 from dagster._core.definitions.asset_check_result import AssetCheckResult
 from dagster._core.definitions.data_version import DataVersion
 from dagster._core.definitions.events import AssetKey, CoercibleToAssetKey
-from dagster._core.definitions.metadata import RawMetadataMapping
-
-
-class AssetResult(
-    NamedTuple(
-        "_AssetResult",
-        [
-            ("asset_key", PublicAttr[Optional[AssetKey]]),
-            ("metadata", PublicAttr[Optional[RawMetadataMapping]]),
-            ("check_results", PublicAttr[Sequence[AssetCheckResult]]),
-            ("data_version", PublicAttr[Optional[DataVersion]]),
-            ("tags", PublicAttr[Optional[Mapping[str, str]]]),
-        ],
-    )
-):
+from dagster._core.definitions.metadata import (  # noqa
+    MetadataValue,
+    RawMetadataMapping,
+    TableSchema,
+)
+from dagster._core.definitions.utils import NoValueSentinel
+
+
+class AssetResult:
```

Comment:
why not just make this a `@record`?

Also I think this is technically a breaking change. Needs to be a record that implements `LegacyNamedTupleMixin`

Reply:
I did originally attempt to make this a record, but the problem is that we need a custom `__new__` for these things, and `@record_custom` does not work with inheritance (i.e. a record that inherits from a record cannot have a custom new method). I recall this being a pretty thorny problem but maybe it's necessary to bulldoze through that before this is approachable (definitely don't think adding `_replace` manually is a mergeable solution, just trying to get tests passing)

Can also see if there's some other way to structure the inheritance that gets around this

Reply:
Ah yes I ran into this too

Reply:
You could write a slightly different variant of `LegacyNamedTupleMixin` meant for use with `dataclass`

Reply:
Here's a cursed dataclass variant: https://gist.github.com/schrockn/00be6a432ff392974c19ef9d0c61b20c